lib/relooper/src/graph/mod.rs

```diff
@@ -1,5 +1,285 @@
+
+    fn lower_edge(&'a self, edge: &'a TE) -> &'a T; //todo rename
+
+    fn edges(&'a self) -> &HashMap<T, Self::EdgeColl>; // change return to Cow?
+    fn nodes(&'a self) -> HashSet<&'a T> {
+        self.edges().keys().collect()
+    }
+
+    fn edge(&'a self, label: &T) -> &'a Self::EdgeColl {
+        self.edges()
+            .get(label)
+            .expect("given node should be present")
+    }
+
+    fn children(&'a self, label: &T) -> HashSet<&'a T> {
+        self.edges()
+            .get(label)
+            .into_iter()
+            .flat_map(|edge_coll| edge_coll.iter().map(|edge| self.lower_edge(edge)))
+            .collect()
+    }
+
+    fn parents(&'a self, label: &T) -> HashSet<&'a T> {
+        self.edges()
+            .iter()
+            .filter_map(|(from, edge_coll)| {
+                edge_coll.iter().find_map(|x| {
+                    if self.lower_edge(x) == label {
+                        Some(from)
+                    } else {
+                        None
+                    }
+                })
+            })
+            .collect()
+    }
+
+    fn reachable(&'a self, node: &'a T) -> HashSet<&T> {
+        Bfs::start_from(node, |x| self.children(x)).collect()
+    }
+
+    fn is_reachable(&'a self, ancestor: &T, descendant: &T) -> bool {
+        let mut descendants = Bfs::start_from(ancestor, |x| self.children(x));
+        descendants.any(|x| x == descendant)
+    }
+
+    // todo rename to 'transpose'?
+    fn in_edges(&'a self) -> HashMap<&'a T, HashSet<&'a T>> {
+        let mut in_edges: HashMap<&T, HashSet<&T>> = HashMap::default();
+
+        for (from, edge_coll) in self.edges() {
+            for to in edge_coll.iter() {
+                in_edges
+                    .entry(self.lower_edge(to))
+                    .or_default()
+                    .insert(from);
+            }
+        }
+
+        in_edges
+    }
+}
+
+pub trait GraphCopy<'a, T: Eq + Hash + Copy + 'a>: Graph<'a, T, T> {
+    fn kosaraju_scc(&'a self) -> Vec<HashSet<T>> {
+        let mut components: Vec<HashSet<T>> = Default::default();
+        let mut visited: HashSet<T> = Default::default();
+
+        let mut transposed = self.in_edges();
+
+        // let mut order: Vec<_> = PrePostOrder::start_from(header, |x| self.children(x))
+        let mut order: Vec<_> = PrePostOrder::start_iter(self.nodes(), |x| self.children(x))
+            .postorder()
+            .collect();
+
+        while let Some(x) = order.pop() {
+            if !visited.contains(x) {
+                let reachable: HashSet<T> =
+                    transposed.reachable(&x).into_iter().map(|&&x| x).collect();
+                for r in reachable.clone() {
+                    visited.insert(r);
+                    transposed.remove_node(&r);
+                }
+                components.push(reachable);
```

Comment:
You could instead write this as

```rust
for r in reachable.iter() {
    transposed.remove_node(r);
    visited.insert(*r);
}
components.push(reachable);
```

I think this will have better performance than calling `clone` on the whole collection because you don't need to allocate memory for all the values an extra time.